### DISCUSSION TITLE
Using custom variables in Vault Lookup Templates?

### DISCUSSION THREAD

**Discussion by jonassvalin** [concourse/concourse#8522] [General]
Hi there,
We have recently set up Vault for secrets management, and we have successfully configured Concourse to be able to pull secrets from it. Works great!

We're now trying to understand how to configure Vault Lookup Templates to be environment aware. For example, in a pipeline for "Service X", we might have a "Deploy Development Blue" task and a "Deploy Development Green" task. We configure these by passing a set of parameters to the task:

```
params:
    DEPLOYMENT_ENVIRONMENT: ((deployment_environment))
    DEPLOYMENT_COLOR: ((deployment_color))
```
          
Is it possible to write the lookup template to use these params? An example would be `/{{.Pipeline}}/{{.DEPLOYMENT_ENVIRONMENT}}/{{.DEPLOYMENT_COLOR}}/{{.Secret}}`.

Or is it only possible to use `Pipeline`, `Team` and `Secret` in templates?
If so, how is it generally recommended to handle environments when dealing with secret lookups?

**Comment by marco-m** [comment]
I am not familiar with Vault lookup templates, but if I understand correctly what you want to achieve, the following simple alternative should work, exploiting the fact that Concourse vars interpolation can be nested.

Granted, if you also need secrets compartmentalisation between environments, then my solution is too simple, although maybe it is possible to do tricks using different teams.

The details are in new section [Two level vars interpolation for parametric secrets](https://github.com/marco-m/concourse-in-a-box/blob/master/doc/concourse-primer.md#two-level-vars-interpolation-for-parametric-secrets) of my [concourse-in-a-box](https://github.com/marco-m/concourse-in-a-box) (all-in-one Concourse based on Docker Compose, with Minio S3-compatible storage and HashiCorp Vault secret manager).

The task

```yaml
    - task: interpolation
      file: concourse-in-a-box/ci/pipelines/05-two-level-interpolation/task.yml
      params:
        ENV: ((env))
        COLOR: ((color))
        SECRET_1: ((secret-1-((env))-((color))))
        SECRET_2: ((secret-2-((env))-((color))))
```

Uses a two level vars interpolation.

For example if we do

```
$ fly -t dev set-pipeline -p interpolation-staging-green \
  -c two-level-interpolation.yml \
  -y env=staging \
  -y color=green
```

We get:

```
      SECRET_1: ((secret-1-staging-green))
      SECRET_2: ((secret-2-staging-green))
```

Once we inject the secrets into Vault:

```
$ vault kv put /concourse/developers/secret-1-staging-green \
  value="I am secret 1 for staging green"
$ vault kv put /concourse/developers/secret-2-staging-green \
  value="I am secret 2 for staging green"
```

Running the job:

```
selected worker: worker-1
...
+ env
...
ENV=staging
COLOR=green
SECRET_1=I am secret 1 for staging green
SECRET_2=I am secret 2 for staging green
```

**Reply by jonassvalin** [reply]
Thank you very much for your comprehensive reply @marco-m ! I think the issue is whether these will be possible to use in the lookup templates, but even if they aren't your suggestion of using nested variable interpolation might still be handy in designing a workaround. Thanks again!
